### PR TITLE
SparseMerkle: Add property tests for internal node

### PR DESCRIPTION
**Pull request by andrewjstone**
New property based tests were added for BatchedInternalNodes using
rapidcheck. These tests generate a bunch of random input and assert that
given invariants hold for the internal structure of the nodes and their
behavior to the outside world.

To facilitate investigation of internal structure, some const member
functions were made public. Furthermore, a `setNibble` function was
created to enable the manipulation of key hashes from tests.

Additionally, the `std::ostream& operator<<(std::ostream&, const T&)`
was fixed for the base types such that the declarations are in the
header file. New versions of these operators were also added to
facilitate debugging.

Lastly, an include guard pragma was added to `util/include/sha3_256.h`
so that it won't generate duplicate declarations and definitions when
included more than once.

### CODE REVIEW

File: kvbc/src/sparse_merkle/keys.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#include "sparse_merkle/keys.h"
+
+namespace concord::storage::sparse_merkle {
+
+std::ostream& operator<<(std::ostream& os, const LeafKey& key) {
+  os << key.hash() << "-" << key.version();
+  return os;
+}
+
+}  // namespace concord::storage::sparse_merkle
```
Comment (dartdart26) [inline kvbc/src/sparse_merkle/keys.cpp]: [Minor]: Missing new line.

File: kvbc/src/sparse_merkle/keys.cpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the
+// LICENSE file.
+
+// All freestanding functions must be in a source file
```
Comment (dartdart26) [inline kvbc/src/sparse_merkle/keys.cpp]: [Minor]: Not sure if we need that comment.

File: kvbc/test/sparse_merkle/internal_node_property_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+};
+
+template <>
+struct Arbitrary<Nibble> {
+  static auto arbitrary() { return gen::construct<Nibble>(gen::inRange<uint8_t>(0, 16)); }
+};
+
+template <>
+struct Arbitrary<Hash> {
+  static auto arbitrary() { return gen::construct<Hash>(gen::arbitrary<SHA3_256::Digest>()); }
```
Comment (dartdart26) [inline kvbc/test/sparse_merkle/internal_node_property_tests.cpp]: Nice!

File: kvbc/test/sparse_merkle/internal_node_property_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+}
+
+RC_GTEST_PROP(batched_internal_node_properties,
+              insert_on_empty_node_contains_leaf_child_at_depth_1,
+              (const LeafChild& child)) {
+  auto node = BatchedInternalNode();
+  auto result = node.insert(child, 0, child.key.version());
+  RC_ASSERT(std::holds_alternative<BatchedInternalNode::InsertComplete>(result));
+
+  // There is no stale leaf overwritte, since we insert a new value to an empty node
```
Comment (dartdart26) [inline kvbc/test/sparse_merkle/internal_node_property_tests.cpp]: [Minor]: Typo.

Comment (dartdart26) [inline kvbc/test/sparse_merkle/internal_node_property_tests.cpp]: [Minor]: Typo.

File: kvbc/test/sparse_merkle/internal_node_property_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    } else {
+      RC_ASSERT(child2.has_value() == false);
+    }
+  }
+}
+
+int main(int argc, char* argv[]) {
+  ::testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```
Comment (dartdart26) [inline kvbc/test/sparse_merkle/internal_node_property_tests.cpp]: [Minor]: Missing new line.

File: kvbc/test/sparse_merkle/internal_node_property_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+  // RC_ASSERT_FALSE does compile with GCC :(
+  RC_ASSERT(std::get<BatchedInternalNode::InsertComplete>(result).stale_leaf.has_value() == false);
+
+  // Only the inserted child exists
+  RC_ASSERT(1u == node.numLeafChildren());
+
+  // The root is the only internal child
+  RC_ASSERT(1u == node.numInternalChildren());
+
+  // The LeafChild is directly under the root
```
Comment (dartdart26) [inline kvbc/test/sparse_merkle/internal_node_property_tests.cpp]: [Minor]: I think we can use `&&` in this particular test as they should both not be internal nodes (either a leaf or empty).

**Comment by dartdart26** [comment]
The `batched_internal_node_properties.inserts_are_deterministic ` test keeps failing on my side. I haven't had the time to investigate why.

File: kvbc/test/sparse_merkle/internal_node_property_tests.cpp (hunk 6 above)
Comment (andrewjstone) [reply kvbc/test/sparse_merkle/internal_node_property_tests.cpp]: Great catch! I wasn't thinking about empty nodes. Will fix.

**Comment by andrewjstone** [comment]
> The `batched_internal_node_properties.inserts_are_deterministic ` test keeps failing on my side. I haven't had the time to investigate why.

It's failing on one of the builds here also. Investigating.

**Comment by andrewjstone** [comment]
> > The `batched_internal_node_properties.inserts_are_deterministic ` test keeps failing on my side. I haven't had the time to investigate why.
> 
> It's failing on one of the builds here also. Investigating.

This should be fixed in the latest commit.

Thanks for the review @dartdart26 !

File: kvbc/test/sparse_merkle/internal_node_property_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+
+  for (auto i = 0u; i < children1.size(); i++) {
+    node1.insert(children1[i], depth, version);
+    node2.insert(children2[i], depth, version);
+  }
+  RC_ASSERT(node1 == node2);
+}
+
+RC_GTEST_PROP(batched_internal_node_properties,
+              any_insert_order_of_unique_keys_maintains_same_keys,
```
Comment (dartdart26) [inline kvbc/test/sparse_merkle/internal_node_property_tests.cpp]: I think that is a good property to test.

**Pull request merged**